### PR TITLE
[AVR-764] add batch transforms

### PR DESCRIPTION
**Pull request by lucabergamini**
Add support for batches in transform points as a new function. This allow to skip a for loop when converting between spaces

### CODE REVIEW

**Comment by lucabergamini** [comment]
@Jimexist FYI

File: l5kit/l5kit/geometry/transform.py [hunk 1]
```diff
@@ -70,6 +70,39 @@ def flip_y_axis(tm: np.ndarray, y_dim_size: int) -> np.ndarray:
     return tm
 
 
+def transform_points_batch(points: np.ndarray, transf_matrix: np.ndarray) -> np.ndarray:
+    """Transform a batch of sequence of points using a batch of transformation matrices.
+    Each sequence is transformed using its own matrix
+    Note this function assumes points.shape[-1] == matrix.shape[11] - 1, which means that last
```
Comment (Jimexist) [inline l5kit/l5kit/geometry/transform.py:76]: `points.shape[-1] == matrix.shape[-1] - 1` not `points.shape[-1] == matrix.shape[11] - 1`

File: l5kit/l5kit/tests/geometry/transform_test.py [hunk 2]
```diff
@@ -2,7 +2,22 @@
 import pytest
 import transforms3d
 
-from l5kit.geometry import transform_point, transform_points
+from l5kit.geometry import transform_point, transform_points, transform_points_batch
```
Comment (Jimexist) [inline l5kit/l5kit/tests/geometry/transform_test.py:5]: I wonder why not call `transform_points_batch` from `transform_points` to reuse code

Comment (lucabergamini) [reply l5kit/l5kit/tests/geometry/transform_test.py:5]: It's an option. The PR is still a draft in the end :)

File: l5kit/l5kit/tests/geometry/transform_test.py [hunk 3]
```diff
@@ -2,7 +2,22 @@
...
+
+
+def test_transform_batch_points() -> None:
+    # transform batch and singular elements one by one, results should match
+    # note: we use random here as the validity of transform is checked below
+
+    tfs = np.random.randn(16, 3, 3)
```
Comment (Jimexist) [inline l5kit/l5kit/tests/geometry/transform_test.py:12]: can you also test if we can jus pass in transform matrix of 2D?

Comment (lucabergamini) [reply l5kit/l5kit/tests/geometry/transform_test.py:12]: > transform matrix of 2D

Do you mean a single matrix for a batch of sequences?

File: l5kit/l5kit/tests/geometry/transform_test.py (hunk 2 above)
Comment (lucabergamini) [reply l5kit/l5kit/tests/geometry/transform_test.py:5]: Now everything calls `transform_points_batch` :) I still need to fix tests to make them really useful (now testing one function against the other makes no sense anymore)

File: l5kit/l5kit/tests/geometry/transform_test.py (hunk 3 above)
Comment (Jimexist) [reply l5kit/l5kit/tests/geometry/transform_test.py:12]: yes, boardcasting basically

File: l5kit/l5kit/geometry/transform.py [hunk 4]
```diff
@@ -70,47 +70,74 @@ def flip_y_axis(tm: np.ndarray, y_dim_size: int) -> np.ndarray:
...
+
+
 def transform_points(points: np.ndarray, transf_matrix: np.ndarray) -> np.ndarray:
     """Transform points using transformation matrix.
     Note this function assumes points.shape[1] == matrix.shape[1] - 1, which means that the last
     row in the matrix does not influence the final result.
     For 2D points only the first 2x3 part of the matrix will be used.
+    This function calls transform_points_batch internally
 
+<<<<<<< HEAD
```
Comment (louis925) [inline l5kit/l5kit/geometry/transform.py:116]: ???

Comment (lucabergamini) [reply l5kit/l5kit/geometry/transform.py:116]: good catch, a leftover from a git merge :)

Comment (louis925) [reply l5kit/l5kit/geometry/transform.py:116]: Looks like the code is not tested

Comment (lucabergamini) [reply l5kit/l5kit/geometry/transform.py:116]: please note it's in a docstring (hence a comment) :)

Comment (louis925) [reply l5kit/l5kit/geometry/transform.py:116]: ahh!

File: l5kit/l5kit/geometry/transform.py [hunk 5]
```diff
@@ -70,47 +70,74 @@ def flip_y_axis(tm: np.ndarray, y_dim_size: int) -> np.ndarray:
     return tm
 
 
+def transform_points_batch(points: np.ndarray, transf_matrix: np.ndarray) -> np.ndarray:
```
Comment (yawei-ye) [inline l5kit/l5kit/geometry/transform.py:73]: @lucabergamini sorry for the super late review, can we combine `transform_points_batch ` with `transform_points`? Just use a single interface

Comment (lucabergamini) [reply l5kit/l5kit/geometry/transform.py]: done :)

File: l5kit/l5kit/geometry/transform.py [hunk 6]
```diff
@@ -71,46 +71,70 @@ def flip_y_axis(tm: np.ndarray, y_dim_size: int) -> np.ndarray:
...
-    """Transform points using transformation matrix.
-    Note this function assumes points.shape[1] == matrix.shape[1] - 1, which means that the last
-    row in the matrix does not influence the final result.
-    For 2D points only the first 2x3 part of the matrix will be used.
+    """
+    Transform points using transformation matrices with 3 modes:
+    - points (N, f), transf_matrix (f+1, f+1)
+        all points are transformed using the matrix and output has shape (N,f).
+    - points (B, N, f), transf_matrix (f+1, f+1)
+        all sequences of points are transformed using the same matrix and output has shape (B, N,f).
```
Comment (yawei-ye) [inline l5kit/l5kit/geometry/transform.py:79]:
```suggestion
        all sequences of points are transformed using the same matrix and the output points have shape (B, N, f).
```

File: l5kit/l5kit/geometry/transform.py [hunk 7]
```diff
@@ -71,46 +71,70 @@ def flip_y_axis(tm: np.ndarray, y_dim_size: int) -> np.ndarray:
...
+        transf_matrix is broadcasted.
+    - points (B, N, f), transf_matrix (B, f+1, f+1)
+        each sequence of points is transformed using its own matrix and output has shape (B, N,f).
```
Comment (yawei-ye) [inline l5kit/l5kit/geometry/transform.py:82]:
```suggestion
        each sequence of points is transformed using its own matrix and the output points have shape (B, N, f).
```

File: l5kit/l5kit/geometry/transform.py [hunk 8]
```diff
@@ -71,46 +71,70 @@ def flip_y_axis(tm: np.ndarray, y_dim_size: int) -> np.ndarray:
...
 
 def transform_points(points: np.ndarray, transf_matrix: np.ndarray) -> np.ndarray:
-    """Transform points using transformation matrix.
-    Note this function assumes points.shape[1] == matrix.shape[1] - 1, which means that the last
-    row in the matrix does not influence the final result.
-    For 2D points only the first 2x3 part of the matrix will be used.
+    """
+    Transform points using transformation matrices with 3 modes:
+    - points (N, f), transf_matrix (f+1, f+1)
+        all points are transformed using the matrix and output has shape (N,f).
```
Comment (yawei-ye) [inline l5kit/l5kit/geometry/transform.py:77]:
```suggestion
        all points are transformed using the matrix and the output points have shape (N, f).
```

File: l5kit/l5kit/geometry/transform.py [hunk 9]
```diff
@@ -71,46 +71,70 @@ def flip_y_axis(tm: np.ndarray, y_dim_size: int) -> np.ndarray:
...
+
+    Note this function assumes points.shape[-1] == matrix.shape[-1] - 1, which means that last
+    rows in the matrices do not influence the final results.
+    For 2D points only the first 2x3 parts of the matrices will be used.
 
     Args:
-        points (np.ndarray): Input points (Nx2) or (Nx3).
-        transf_matrix (np.ndarray): 3x3 or 4x4 transformation matrix for 2D and 3D input respectively
+        points (np.ndarray): Input points (N, 2), (N, 3, (BxNx2), (BxNx3).
+        transf_matrix (np.ndarray): Transformation matrix (3x3), (4x4), (Bx3x3), (Bx4x4).
```
Comment (yawei-ye) [inline l5kit/l5kit/geometry/transform.py:90]:
```suggestion
        transf_matrix (np.ndarray): Transformation matrix of shape (f+1, f+1) or (B, f+1, f+1) with f = 2 or 3.
```

File: l5kit/l5kit/geometry/transform.py (hunk 8 above)
Comment (yawei-ye) [reply l5kit/l5kit/geometry/transform.py:77]: @lucabergamini does `f` stands for `feature`? Should we use  capitalized form `F`?

Comment (yawei-ye) [reply l5kit/l5kit/geometry/transform.py]: Ping, should we use `F` for the last feature dimension?

Comment (yawei-ye) [reply l5kit/l5kit/geometry/transform.py]: @lucabergamini

**Review by yawei-ye — APPROVED** [review]
LGTM!

**Review by yawei-ye — APPROVED** [review]

**Pull request merged**